File: app/lib/domain/usecases/generate_session.dart
import '../entities/concept.dart';
import '../entities/review_card.dart';
import '../entities/quiz_question.dart';
import '../entities/session_config.dart';

/// A generated daily session with cards organized by phase.
class GeneratedSession {
  final List<ReviewCard> warmupCards;
  final List<ReviewCard> coreCards;
  final List<QuizQuestion> challengeQuestions;
  final SessionConfig config;

  const GeneratedSession({
    required this.warmupCards,
    required this.coreCards,
    required this.challengeQuestions,
    required this.config,
  });

  int get totalCards => warmupCards.length + coreCards.length;
  int get totalQuestions => challengeQuestions.length;
}

/// Generates a structured daily session from available review cards and quizzes.
class GenerateSession {
  const GenerateSession();

  GeneratedSession call({
    required List<ReviewCard> dueCards,
    required List<ReviewCard> easyCards,
    required List<QuizQuestion> availableQuizzes,
    required Map<String, Concept> conceptMap,
    SessionConfig config = const SessionConfig(),
  }) {
    // Phase 1: Warmup — pick easy cards the user knows well
    final warmup = _selectWarmupCards(easyCards, config.warmupCards);

    // Phase 2: Core — due review cards, interleaved across domains
    final core = _selectCoreCards(
      dueCards,
      conceptMap,
      config.coreReviewCards,
    );

    // Phase 3: Challenge — quiz questions related to today's reviewed concepts
    final reviewedConceptIds = {
      ...warmup.map((c) => c.conceptId),
      ...core.map((c) => c.conceptId),
    };

    final challenge = _selectChallengeQuestions(
      availableQuizzes,
      reviewedConceptIds,
      config.challengeQuestions,
    );

    return GeneratedSession(
      warmupCards: warmup,
      coreCards: core,
      challengeQuestions: challenge,
      config: config,
    );
  }

  List<ReviewCard> _selectWarmupCards(
    List<ReviewCard> easyCards,
    int count,
  ) {
    // Sort by highest ease factor (strongest knowledge)
    final sorted = [...easyCards]
      ..sort((a, b) => b.easeFactor.compareTo(a.easeFactor));
    return sorted.take(count).toList();
  }

  List<ReviewCard> _selectCoreCards(
    List<ReviewCard> dueCards,
    Map<String, Concept> conceptMap,
    int maxCount,
  ) {
    if (dueCards.length <= maxCount) return dueCards;

    // Interleave across domains for better retention
    final byDomain = <String, List<ReviewCard>>{};
    for (final card in dueCards) {
      final concept = conceptMap[card.conceptId];
      final domain =
          (concept != null && concept.tags.isNotEmpty) ? concept.tags.first : 'unknown';
      byDomain.putIfAbsent(domain, () => <ReviewCard>[]).add(card);
    }

    final result = <ReviewCard>[];
    var added = true;
    while (result.length < maxCount && added) {
      added = false;
      for (final domain in byDomain.keys) {
        if (result.length >= maxCount) break;
        final cards = byDomain[domain]!;
        if (cards.isNotEmpty) {
          result.add(cards.removeAt(0));
          added = true;
        }
      }
    }

    return result;
  }

  List<QuizQuestion> _selectChallengeQuestions(
    List<QuizQuestion> available,
    Set<String> reviewedConceptIds,
    int count,
  ) {
    // Prefer questions related to today's reviewed concepts
    final related = available.where((q) {
      return q.conceptIds.any((id) => reviewedConceptIds.contains(id));
    }).toList();

    final unrelated = available.where((q) {
      return !q.conceptIds.any((id) => reviewedConceptIds.contains(id));
    }).toList();

    final result = <QuizQuestion>[];
    result.addAll(related.take(count));
    if (result.length < count) {
      result.addAll(unrelated.take(count - result.length));
    }

    return result;
  }
}

File: web/src/lib/engine/generate-session.ts
/**
 * Daily session generator. Port of
 * `app/lib/domain/usecases/generate_session.dart`: warmup (strongest cards),
 * core (due cards interleaved across domains), challenge (quiz questions
 * biased toward today's reviewed concepts).
 */

export type SessionCard = {
  conceptId: string;
  easeFactor: number;
};

export type SessionQuizQuestion = {
  id: string;
  concept_ids: string[];
};

export type SessionConfig = {
  warmupCards: number;
  coreReviewCards: number;
  challengeQuestions: number;
};

export type GeneratedSession<
  TCard extends SessionCard,
  TQuestion extends SessionQuizQuestion,
> = {
  warmupCards: TCard[];
  coreCards: TCard[];
  challengeQuestions: TQuestion[];
  config: SessionConfig;
};

export function generateSession<
  TCard extends SessionCard,
  TQuestion extends SessionQuizQuestion,
>({
  dueCards,
  easyCards,
  availableQuestions,
  domainOf,
  config,
}: {
  dueCards: TCard[];
  easyCards: TCard[];
  availableQuestions: TQuestion[];
  /** Maps a conceptId to its domain, for interleaving. */
  domainOf: (conceptId: string) => string;
  config: SessionConfig;
}): GeneratedSession<TCard, TQuestion> {
  const warmup = selectWarmup(easyCards, config.warmupCards);
  const core = selectCore(dueCards, domainOf, config.coreReviewCards);

  const reviewedIds = new Set([
    ...warmup.map((c) => c.conceptId),
    ...core.map((c) => c.conceptId),
  ]);
  const challenge = selectChallenge(
    availableQuestions,
    reviewedIds,
    config.challengeQuestions,
  );

  return { warmupCards: warmup, coreCards: core, challengeQuestions: challenge, config };
}

function selectWarmup<TCard extends SessionCard>(
  easyCards: TCard[],
  count: number,
): TCard[] {
  return [...easyCards]
    .sort((a, b) => b.easeFactor - a.easeFactor)
    .slice(0, count);
}

function selectCore<TCard extends SessionCard>(
  dueCards: TCard[],
  domainOf: (conceptId: string) => string,
  maxCount: number,
): TCard[] {
  if (dueCards.length <= maxCount) return dueCards;

  const byDomain = new Map<string, TCard[]>();
  for (const card of dueCards) {
    const domain = domainOf(card.conceptId);
    const list = byDomain.get(domain) ?? [];
    list.push(card);
    byDomain.set(domain, list);
  }

  const result: TCard[] = [];
  let added = true;
  while (result.length < maxCount && added) {
    added = false;
    for (const cards of byDomain.values()) {
      if (result.length >= maxCount) break;
      const next = cards.shift();
      if (next) {
        result.push(next);
        added = true;
      }
    }
  }
  return result;
}

function selectChallenge<TQuestion extends SessionQuizQuestion>(
  available: TQuestion[],
  reviewedConceptIds: Set<string>,
  count: number,
): TQuestion[] {
  const related = available.filter((q) =>
    q.concept_ids.some((id) => reviewedConceptIds.has(id)),
  );
  const unrelated = available.filter(
    (q) => !q.concept_ids.some((id) => reviewedConceptIds.has(id)),
  );
  const result = related.slice(0, count);
  if (result.length < count) {
    result.push(...unrelated.slice(0, count - result.length));
  }
  return result;
}
